test/fixtures/azure-fixture/src/main/java/fixture/azure/MockAzureBlobStore.java

```diff
@@ -0,0 +1,483 @@
+                mockAzureBlockBlob.putBlock(blockId, content, leaseId);
+                return mockAzureBlockBlob;
+            }
+        });
+    }
+
+    public void putBlockList(String path, List<String> blockIds, @Nullable String leaseId) {
+        final AzureBlob azureBlob = getExistingBlob(path);
+        if (azureBlob instanceof MockAzureBlockBlob mabb) {
+            mabb.putBlockList(blockIds, leaseId);
+        } else {
+            throw new ConflictException("InvalidBlobType", "The blob type is invalid for this operation.");
+        }
+    }
+
+    public void putBlob(String path, BytesReference contents, BlobType blobType, @Nullable String ifNoneMatch, @Nullable String leaseId) {
+        blobs.compute(path, (p, existingValue) -> {
+            if (existingValue != null) {
+                existingValue.setContents(contents, leaseId, ifNoneMatch);
+                return existingValue;
+            } else {
+                final AzureBlob newBlob = blobType.create();
+                newBlob.setContents(contents, leaseId);
+                return newBlob;
+            }
+        });
+    }
+
+    public AzureBlob getBlob(String path, @Nullable String leaseId) {
+        final AzureBlob azureBlob = getExistingBlob(path);
+        azureBlob.assertCanRead(leaseId);
+        return azureBlob;
+    }
+
+    public void deleteBlob(String path, @Nullable String leaseId) {
+        final AzureBlob azureBlob = getExistingBlob(path);
+        azureBlob.assertCanWrite(leaseId);
+        blobs.remove(path);
+    }
+
+    public Map<String, AzureBlob> listBlobs(String prefix, @Nullable String leaseId) {
+        return blobs.entrySet().stream().filter(e -> {
+            if (prefix == null || e.getKey().startsWith(prefix)) {
+                return true;
+            }
+            return false;
+        })
+            .filter(e -> e.getValue().isCommitted())
+            .peek(e -> e.getValue().assertCanRead(leaseId))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    public String acquireLease(String path, int leaseTimeSeconds, @Nullable String proposedLeaseId) {
+        final AzureBlob azureBlob = getExistingBlob(path);
+        return azureBlob.acquireLease(proposedLeaseId, leaseTimeSeconds);
+    }
+
+    public void releaseLease(String path, @Nullable String leaseId) {
+        final AzureBlob azureBlob = getExistingBlob(path);
+        azureBlob.releaseLease(leaseId);
+    }
+
+    public Map<String, BytesReference> blobs() {
+        return Maps.transformValues(blobs, AzureBlob::getContents);
+    }
+
+    private AzureBlob getExistingBlob(String path) {
+        final AzureBlob azureBlob = blobs.get(path);
+        if (azureBlob == null) {
+            throw new NotFoundException("BlobNotFound", "The specified blob does not exist.");
+        }
+        return azureBlob;
+    }
+
+    public interface AzureBlob {
```

Comment:
Do we really need an `interface AzureBlob` and an `abstract class AbstractAzureBlob implements AzureBlob` and then a concrete `class MockAzureBlockBlob extends AbstractAzureBlob`? There's only one concrete class in play here, the extra layers of abstraction seem to make this harder to follow.

Reply:
I flattened the hierarchy in 2b21457fdebab11629796a9f420e0f897200b828. I think keeping the interface and the single implementation is worthwhile because the `AzureHttpHandler` shouldn't be doing some of the things that `MockAzureBlockBlob` implements. So the interface is the "public" interface, and also the non-type-specific interface.

Reply:
> the AzureHttpHandler shouldn't be doing some of the things that MockAzureBlockBlob implements

Sorta, but then again you can bypass that by casting the blob to its concrete type, and indeed we do exactly that already to work around the very restriction you're promoting 😁 Let's just use the concrete type everywhere.

Reply:
The implementation class was private, so you couldn't cast on that side of the interface. But I removed it in 0b382d19641ceffe5f9219c8a0413365b089d9db anyhow.